src/main/java/com/aws/greengrass/lifecyclemanager/KernelLifecycle.java

```diff
@@ -592,4 +594,31 @@ public void shutdown(int timeoutSeconds) {
     GreengrassService getMain() {
         return mainService;
     }
+
+    /**
+     * Check if all services has reached to terminal state: RUNNING, FINISHED or BROKEN.
+     *
+     */
+    public boolean allServicesInTerminalState() {
+        //Check all other component status, false if any component is not at the terminal state
+        List<GreengrassService> servicesToTrack = kernel.orderedDependencies().stream()
+                .filter(GreengrassService::shouldAutoStart).collect(Collectors.toList());
+        boolean allReachedTerminalState = true;
+        for (GreengrassService service : servicesToTrack) {
+            State state = service.getState();
+            //Consider RUNNING, FINISHED and BROKEN as terminal state, as component reached to the end of lifecycle
+            if (State.BROKEN.equals(state)) {
+                continue;
+            }
+            if (!service.reachedDesiredState()) {
+                allReachedTerminalState = false;
+                break;
+            }
+            if (State.RUNNING.equals(state) || State.FINISHED.equals(state)) {
```

Comment:
nit: you can combine this with condition on 610

Reply:
But broken is not entering desired state, is that more confusing to write it as `if((reachedDesiredstate && RUNNING || FINISHED) || BROKEN)`

Reply:
I am not suggesting to club all the conditions together, only club conditions 1 and 3 since they are repetitive

Reply:
Updated